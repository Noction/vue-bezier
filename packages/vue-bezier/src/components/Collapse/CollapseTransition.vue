<template>
  <component
    :is="componentType"
    :tag="tag"
    v-bind="{ ...$attrs, ...hooks }"
    move-class="collapse-move"
  >
    <slot />
  </component>
</template>

<script setup lang="ts">
import { BaseTransitionProps } from 'vue'
import {
  buildComponentType,
  buildEmits,
  buildProps,
  buildTag
} from '../../composable'
import { leave, setAbsolutePosition, setStyles } from '../../composable/buildHooks'

const props = defineProps(buildProps())
const emit = defineEmits(buildEmits())

const componentType = buildComponentType(props)
const tag = buildTag(props)

const hooks: BaseTransitionProps = {
  onAfterEnter (el) {
    // for safari: remove class then reset height is necessary
    el.style.transition = ''
    el.style.height = ''
    el.style.overflow = el.dataset.oldOverflow

    emit('after-enter', el)
  },
  onAfterLeave (el) {
    el.style.transition = ''
    el.style.height = ''
    el.style.overflow = el.dataset.oldOverflow
    el.style.paddingTop = el.dataset.oldPaddingTop
    el.style.paddingBottom = el.dataset.oldPaddingBottom

    emit('after-leave', el)
  },
  onBeforeEnter (el) {
    const enterDuration = props.duration?.enter ?? props.duration ?? 0

    el.style.transition = transitionStyle(enterDuration)

    if (!el.dataset) el.dataset = {}

    el.dataset.oldPaddingTop = el.style.paddingTop
    el.dataset.oldPaddingBottom = el.style.paddingBottom

    el.style.height = '0'
    el.style.paddingTop = 0
    el.style.paddingBottom = 0

    setStyles(props, el)

    emit('before-enter', el)
  },
  onBeforeLeave (el) {
    if (!el.dataset) el.dataset = {}

    el.dataset.oldPaddingTop = el.style.paddingTop
    el.dataset.oldPaddingBottom = el.style.paddingBottom
    el.dataset.oldOverflow = el.style.overflow

    el.style.height = `${el.scrollHeight}px`
    el.style.overflow = 'hidden'

    setStyles(props, el)

    emit('before-leave', el)
  },
  onEnter (el) {
    el.dataset.oldOverflow = el.style.overflow

    if (el.scrollHeight !== 0) {
      el.style.height = `${el.scrollHeight}px`
      el.style.paddingTop = el.dataset.oldPaddingTop
      el.style.paddingBottom = el.dataset.oldPaddingBottom
    } else {
      el.style.height = ''
      el.style.paddingTop = el.dataset.oldPaddingTop
      el.style.paddingBottom = el.dataset.oldPaddingBottom
    }

    el.style.overflow = 'hidden'
  },
  onLeave (el, done: () => void) {
    const leaveDuration = props.duration.leave ?? props.duration ?? 0

    if (el.scrollHeight !== 0) {
      // for safari: add class after set height, or it will jump to zero height suddenly, weired
      el.style.transition = transitionStyle(leaveDuration)
      el.style.height = 0
      el.style.paddingTop = 0
      el.style.paddingBottom = 0
    }

    // necessary for transition-group
    setAbsolutePosition(props, el)

    leave(props, el, done)
    emit('leave', el, done)
  }
}

const transitionStyle = (duration: number) => {
  const durationInSeconds = duration / 1000

  return `${durationInSeconds}s height ease-in-out, ${durationInSeconds}s padding-top ease-in-out, ${durationInSeconds}s padding-bottom ease-in-out`
}

</script>

<style>
  .collapse-move { transition: transform .3s ease-in-out; }
</style>
