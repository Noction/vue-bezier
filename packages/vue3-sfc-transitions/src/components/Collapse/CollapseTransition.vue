<template>
  <component
    :is="componentType"
    :tag="tag"
    v-bind="{ ...$attrs, ...hooks }"
    move-class="collapse-move"
  >
    <slot />
  </component>
</template>

<script setup lang="ts">
import { BaseTransitionProps } from 'vue'
import {
  buildComponentType,
  buildEmits,
  buildProps,
  buildTag
} from '../../composable'
import { setAbsolutePosition, setStyles } from '../../composable/buildHooks'

const props = defineProps(buildProps())
const emit = defineEmits(buildEmits())

const componentType = buildComponentType(props)
const tag = buildTag(props)

const hooks: BaseTransitionProps = {
  onAfterEnter (el) {
    // for safari: remove class then reset height is necessary
    el.style.transition = ''
    el.style.height = ''
    el.style.overflow = el.dataset.oldOverflow
  },
  onAfterLeave (el) {
    el.style.transition = ''
    el.style.height = ''
    el.style.overflow = el.dataset.oldOverflow
    el.style.paddingTop = el.dataset.oldPaddingTop
    el.style.paddingBottom = el.dataset.oldPaddingBottom
  },
  onBeforeEnter (el) {
    const enterDuration = props.duration?.enter ?? props.duration ?? 0

    el.style.transition = transitionStyle(enterDuration)

    if (!el.dataset) el.dataset = {}

    el.dataset.oldPaddingTop = el.style.paddingTop
    el.dataset.oldPaddingBottom = el.style.paddingBottom

    el.style.height = '0'
    el.style.paddingTop = 0
    el.style.paddingBottom = 0

    setStyles(props, el)
  },
  onBeforeLeave (el) {
    if (!el.dataset) el.dataset = {}

    el.dataset.oldPaddingTop = el.style.paddingTop
    el.dataset.oldPaddingBottom = el.style.paddingBottom
    el.dataset.oldOverflow = el.style.overflow

    el.style.height = `${el.scrollHeight}px`
    el.style.overflow = 'hidden'

    setStyles(props, el)
  },
  onEnter (el) {
    el.dataset.oldOverflow = el.style.overflow

    if (el.scrollHeight !== 0) {
      el.style.height = `${el.scrollHeight}px`
      el.style.paddingTop = el.dataset.oldPaddingTop
      el.style.paddingBottom = el.dataset.oldPaddingBottom
    } else {
      el.style.height = ''
      el.style.paddingTop = el.dataset.oldPaddingTop
      el.style.paddingBottom = el.dataset.oldPaddingBottom
    }

    el.style.overflow = 'hidden'
  },
  onLeave (el) {
    const leaveDuration = props.duration.leave ?? props.duration ?? 0

    if (el.scrollHeight !== 0) {
      // for safari: add class after set height, or it will jump to zero height suddenly, weired
      el.style.transition = transitionStyle(leaveDuration)
      el.style.height = 0
      el.style.paddingTop = 0
      el.style.paddingBottom = 0
    }

    // necessary for transition-group
    setAbsolutePosition(props, el)
  }
}

const transitionStyle = (duration = 300) => {
  const durationInSeconds = duration / 1000

  return  `${durationInSeconds}s height ease-in-out, ${durationInSeconds}s padding-top ease-in-out, ${durationInSeconds}s padding-bottom ease-in-out`
}

</script>

<style>
  .collapse-move { transition: transform .3s ease-in-out; }
</style>
