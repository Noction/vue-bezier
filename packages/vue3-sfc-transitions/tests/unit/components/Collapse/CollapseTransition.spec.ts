import CollapseTransition from '../../../../src/components/Collapse/CollapseTransition.vue'
import { mount } from '@vue/test-utils'

describe('components / Collapse / CollapseTransition', () => {
  it('When mounting component with default props, it should work fine', () => {
    const wrapper = mount(CollapseTransition)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('When mounting component with group enabled prop, it should work fine', () => {
    const wrapper = mount(CollapseTransition, { props: { group: true } })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
