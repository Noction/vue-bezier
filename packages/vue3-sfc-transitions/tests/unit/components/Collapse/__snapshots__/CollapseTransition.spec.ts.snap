// Vitest Snapshot v1

exports[`components / Collapse / CollapseTransition > When mounting component with default props, it should work fine 1`] = `"<transition-stub appear=\\"false\\" persisted=\\"false\\" css=\\"true\\" tag=\\"span\\" move-class=\\"collapse-move\\"></transition-stub>"`;

exports[`components / Collapse / CollapseTransition > When mounting component with group enabled prop, it should work fine 1`] = `"<transition-group-stub tag=\\"span\\" moveclass=\\"collapse-move\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\"></transition-group-stub>"`;
